### PR TITLE
virtio-queue: improve testing

### PR DESCRIPTION
**Pull request by lauralt**
This PR adds or improves tests for validating the notification suppression mechanism, the reset, indirect descriptor requirements and processing of them.

### CODE REVIEW

**Review by jiangliu — DISMISSED** [review]

File: crates/virtio-queue/src/lib.rs [hunk 1]
```diff
@@ -1182,12 +1182,38 @@ mod tests {
...
+            assert_eq!(c.ttl, 0);
         }
     }
 
+    #[test]
+    fn test_ttl_wrap_around() {
+        let m = &GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x100000)]).unwrap();
+        let vq = MockSplitQueue::new(m, 16);
+
+        for j in 0..15 {
```
Comment (andreeaflorescu) [inline crates/virtio-queue/src/lib.rs:1200]: nit: Why use j?

File: crates/virtio-queue/src/lib.rs [hunk 2]
```diff
@@ -1182,12 +1182,38 @@ mod tests {
...
             assert!(c.next().is_some());
+            // The descriptor above was the last from the chain, so `ttl` should be 0 now.
+            assert_eq!(c.ttl, 0);
             assert!(c.next().is_none());
+            assert_eq!(c.ttl, 0);
         }
     }
 
+    #[test]
+    fn test_ttl_wrap_around() {
```
Comment (andreeaflorescu) [inline crates/virtio-queue/src/lib.rs:1196]: I am not sure why is this called `wrap_around`. Can you add some more details about the test case that you're adding here?

File: crates/virtio-queue/src/lib.rs [hunk 3]
```diff
@@ -1493,11 +1567,35 @@ mod tests {
...
+        q.state.signalled_used = Some(Wrapping(15));
+        assert_eq!(q.state.size, 8);
+        // `create_queue` also marks the queue as ready.
+        assert_eq!(q.state.ready, true);
+        assert_ne!(q.state.desc_table, GuestAddress(DEFAULT_DESC_TABLE_ADDR));
+        assert_ne!(q.state.avail_ring, GuestAddress(DEFAULT_AVAIL_RING_ADDR));
+        assert_ne!(q.state.used_ring, GuestAddress(DEFAULT_USED_RING_ADDR));
+        assert_ne!(q.state.next_avail, Wrapping(0));
+        assert_ne!(q.state.next_used, Wrapping(0));
+        assert_ne!(q.state.signalled_used, None);
```
Comment (andreeaflorescu) [inline crates/virtio-queue/src/lib.rs:1587]: Why do we check these before and after the reset? They seem to have the same values.

File: crates/virtio-queue/src/lib.rs [hunk 4]
```diff
@@ -1544,6 +1642,11 @@ mod tests {
         q.state.next_used = Wrapping(0);
         assert_eq!(q.needs_notification().unwrap(), true);
         assert_eq!(q.needs_notification().unwrap(), false);
+
+        m.write_obj::<u16>(u16::MAX - 3, avail_addr.unchecked_add(4 + qsize as u64 * 2))
```
Comment (andreeaflorescu) [inline crates/virtio-queue/src/lib.rs:1646]: Can you add a comment here to explain why you're calling write again & needs notification is true?

File: crates/virtio-queue/src/lib.rs [hunk 5]
```diff
@@ -1582,4 +1685,154 @@ mod tests {
...
+
+        let mut i = 0;
+
+        loop {
+            i += 1;
+            q.disable_notification().unwrap();
+
+            while let Some(_chain) = q.iter().unwrap().next() {
+                // Do something with the chain. Here the device would consume entries from the
+                // available ring, add an entry in the used ring and optionally notify the driver.
```
Comment (andreeaflorescu) [inline crates/virtio-queue/src/lib.rs:1728]: I would also add a comment that for the purpose of this test what happens with the chain is not relevant, so we just ignore it as we just want to consume the entries in the queue.

**Review by andreeaflorescu — COMMENTED** [review]
Can you add some more comments in the tests? For some of them it's not obvious why we're adding them, so maybe we can give a short description at the beginning (i.e. this is a negative test, its purpose it to check that in case of  misconfiguration (name the misconfiguration) the device does not cause a panic).

Can you also update the commit message? I think you were saying that some of the tests that we already have were not following the specification. We should name them in the commit message to have a history. We should also say in the commit message what kind of tests we're adding and why. What is particularly interesting about the tests is that we are adding negative tests checking what happens when the driver is not following the spec (i.e. the things with MUST).

Just saying in the commit message that we're "improving testing" is an understatement ;)

File: crates/virtio-queue/src/lib.rs (hunk 2 above)
Comment (lauralt) [reply crates/virtio-queue/src/lib.rs:1196]: Added some comments in the test.

File: crates/virtio-queue/src/lib.rs (hunk 1 above)
Comment (lauralt) [reply crates/virtio-queue/src/lib.rs:1200]: Just some copy paste from another test, I changed it with `i`.

File: crates/virtio-queue/src/lib.rs (hunk 3 above)
Comment (lauralt) [reply crates/virtio-queue/src/lib.rs:1587]: I am using `assert_ne!` before the reset.

File: crates/virtio-queue/src/lib.rs (hunk 4 above)
Comment (lauralt) [reply crates/virtio-queue/src/lib.rs:1646]: done

File: crates/virtio-queue/src/lib.rs (hunk 5 above)
Comment (lauralt) [reply crates/virtio-queue/src/lib.rs:1728]: done

**Comment by lauralt** [comment]
> Can you add some more comments in the tests? For some of them it's not obvious why we're adding them, so maybe we can give a short description at the beginning (i.e. this is a negative test, its purpose it to check that in case of misconfiguration (name the misconfiguration) the device does not cause a panic).

I added more comments.

> 
> Can you also update the commit message? I think you were saying that some of the tests that we already have were not following the specification. We should name them in the commit message to have a history. We should also say in the commit message what kind of tests we're adding and why. What is particularly interesting about the tests is that we are adding negative tests checking what happens when the driver is not following the spec (i.e. the things with MUST).

I updated the commit message. I'm not fixing yet the tests that were not following the spec. This was part of [this issue](https://github.com/rust-vmm/vm-virtio/issues/95) and I'll handle it in a separate PR.

> 
> Just saying in the commit message that we're "improving testing" is an understatement ;)

**Comment by lauralt** [comment]
> > Can you add some more comments in the tests? For some of them it's not obvious why we're adding them, so maybe we can give a short description at the beginning (i.e. this is a negative test, its purpose it to check that in case of misconfiguration (name the misconfiguration) the device does not cause a panic).
> 
> I added more comments.
> 
> > Can you also update the commit message? I think you were saying that some of the tests that we already have were not following the specification. We should name them in the commit message to have a history. We should also say in the commit message what kind of tests we're adding and why. What is particularly interesting about the tests is that we are adding negative tests checking what happens when the driver is not following the spec (i.e. the things with MUST).
> 
> I updated the commit message. I'm not fixing yet the tests that were not following the spec. This was part of [this issue](https://github.com/rust-vmm/vm-virtio/issues/95) and I'll handle it in a separate PR.
> 
> > Just saying in the commit message that we're "improving testing" is an understatement ;)

I added a separate commit for https://github.com/rust-vmm/vm-virtio/issues/95 and also did some fixes to the already existing test for indirect descriptors. More details in the commit message/comments.

**Review by slp — APPROVED** [review]
LGTM, thanks!

**Review by jiangliu — APPROVED** [review]

**Pull request merged**